scvi/external/cell2location/_model.py

```diff
@@ -0,0 +1,572 @@
+        
+
+    def train_pyro_v2(self, max_epochs: Optional[int] = None,
+                   use_gpu: Optional[bool] = None,
+                   train_size: float = 1,
+                   validation_size: Optional[float] = None,
+                   batch_size: Optional[int] = None, lr: float = 0.001,
+                   total_grad_norm_constraint: float = 200,
+                   **kwargs):
+        
+        if use_gpu is True:
+            self.module.cuda()
+        
+        plan_kwargs = {'loss_fn': pyro.infer.Trace_ELBO(), # for some reason JitTrace_ELBO does not work
+                       'optim': pyro.optim.ClippedAdam({'lr': lr,
+                                                        # limit the gradient step from becoming too large
+                                                        'clip_norm': total_grad_norm_constraint})}
+        
+        if batch_size is None:
+            batch_size = self.batch_size
+        
+        self.train(max_epochs=max_epochs, use_gpu=use_gpu,
+                   train_size=train_size, validation_size=validation_size, batch_size=batch_size,
+                   plan_kwargs=plan_kwargs,
+                   plan_class=PyroTrainingPlan)
+    
+    
+    def train_pyro(
+        self,
+        max_epochs: int = 30000,
+        lr: float = 0.001,
+        total_grad_norm_constraint: float = 200,
+        use_gpu: Optional[bool] = None,
+        train_size: float = 1,
+        validation_size: Optional[float] = None,
+        batch_size = None,
+        plan_kwargs: Optional[dict] = None,
+        **kwargs,
+    ):
+        """
+        Train the model using variational inference (Pyro).
+
+        Parameters
+        ----------
+        max_epochs
+            Number of epochs to train for
+        lr
+            Learning rate for optimization.
+        use_gpu
+            If `True`, use the GPU if available.
+        train_size
+            Size of training set in the range [0.0, 1.0].
+        validation_size
+            Size of the test set. If `None`, defaults to 1 - `train_size`. If
+            `train_size + validation_size < 1`, the remaining cells belong to a test set.
+        batch_size
+            Minibatch size to use during training.
+        **kwargs
+            Other keyword args for :class:`~scvi.lightning.Trainer`.
+        """
+        
+        if use_gpu is None:
+            use_gpu = self.use_gpu
+        else:
+            use_gpu = use_gpu and torch.cuda.is_available()
+        gpus = 1 if use_gpu else None
+        
+        if self.use_gpu:
+            torch.set_default_tensor_type(torch.cuda.FloatTensor)
+        else:
+            torch.set_default_tensor_type(torch.FloatTensor)
+        
+        
+        if use_gpu:
+            self.module.cuda()
+        
+        train_dl = AnnDataLoader(self.adata, shuffle=True, 
+                                 batch_size=self.batch_size, 
+                                 drop_last=True)
+        pyro.clear_param_store()
```

Comment:
I do not know if it is possible (or a good idea) to change pyro plate size on the fly but otherwise would be good to have `drop_last` in the training plan.